modules/mixminus/mixminus.c

```diff
@@ -0,0 +1,401 @@
+		enc = le->data;
+		if (!enc)
+			continue;
+		if (enc->au == au)
+			continue;
+
+		mix = mem_zalloc(sizeof(*mix), mix_destructor);
+		if (!mix)
+			return ENOMEM;
+
+		err = aubuf_alloc(&mix->ab, psize, 3 * psize);
+		if (err)
+			return err;
+
+		mix->au = st->au; /* using audio object as id */
+		mix->srate = st->srate;
+		mix->ch = st->ch;
+		list_append(&enc->mixers, &mix->le_priv, mix);
+	}
+
+	/* add other mixes to new enc */
+	for (le = list_head(&encs); le; le = le->next) {
+		enc = le->data;
+		if (!enc)
+			continue;
+		if (enc->au == au)
+			continue;
+
+		mix = mem_zalloc(sizeof(*mix), mix_destructor);
+		if (!mix)
+			return ENOMEM;
+
+		err = aubuf_alloc(&mix->ab, psize, 3 * psize);
+		if (err)
+			return err;
+
+		mix->au = enc->au; /* using audio object as id */
+		mix->srate = enc->srate;
+		mix->ch = enc->ch;
+		list_append(&st->mixers, &mix->le_priv, mix);
+	}
+
+	*stp = (struct aufilt_enc_st *) st;
+
+	return 0;
+}
+
+
+static int decode_update(struct aufilt_dec_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct mixminus_dec *st;
+	(void)af;
+	(void)prm;
+
+	if (!stp || !ctx)
+		return EINVAL;
+
+	if (prm->fmt != AUFMT_S16LE) {
+		warning("mixminus: Only S16LE is supported.\n");
+		return EINVAL;
+	}
+
+	if (*stp)
+		return 0;
+
+	st = mem_zalloc(sizeof(*st), NULL);
+	if (!st)
+		return ENOMEM;
+
+	st->au = au;
+
+	*stp = (struct aufilt_dec_st *)st;
+
+	return 0;
+}
+
+
+static int encode(struct aufilt_enc_st *aufilt_enc_st, struct auframe *af)
+{
+	struct mixminus_enc *enc = (struct mixminus_enc *)aufilt_enc_st;
+	size_t i, inc;
+	size_t outc;
+	struct le *lem;
+	struct mix *mix;
+	int16_t *sampv = af->sampv;
+	int16_t *sampv_mix = enc->sampv;
+	int err = 0;
+
+	for (lem = list_head(&enc->mixers); lem; lem = lem->next) {
+		mix = lem->data;
+		if (!mix)
+			continue;
+
+		if (!audio_is_conference(mix->au))
+			continue;
+
+		err = auresamp_setup(&enc->resamp, mix->srate, mix->ch,
+				     enc->srate, enc->ch);
+		if (err) {
+			warning("mixminus/auresamp_setup error (%m)\n", err);
+			return err;
+		}
+
+		if (enc->resamp.resample) {
+			outc = AUDIO_SAMPSZ;
+			sampv_mix = enc->rsampv;
+
+			if (enc->srate > mix->srate) {
+				inc = af->sampc / enc->resamp.ratio;
+			}
+			else {
+				inc = af->sampc * enc->resamp.ratio;
+			}
+
+			if (enc->ch == 2 && mix->ch == 1) {
+				inc = inc / 2;
+			}
+
+			if (enc->ch == 1 && mix->ch == 2) {
+				inc = inc * 2;
+			}
+
+			aubuf_read_samp(mix->ab, enc->sampv, inc);
+
+			err = auresamp(&enc->resamp, sampv_mix, &outc,
+				       enc->sampv, inc);
+			if (err) {
+				warning("mixminus/auresamp error (%m)\n", err);
+				return err;
+			}
+			if (outc != af->sampc) {
+				warning("mixminus/auresamp sample count "
+					"error\n");
+				return EINVAL;
+			}
+		}
+		else {
+			aubuf_read_samp(mix->ab, sampv_mix, af->sampc);
+		}
+
+		for (i = 0; i < af->sampc; i++) {
+			sampv[i] = sampv[i] + sampv_mix[i];
+		}
```

Comment:
did you consider saturation of added samples here ?

the range is 16-bit signed, if two added samples have e.g. value 32000 and 32000,
the sum will be 64000 and outside the range.

it is possible to add the samples using a 32-bit integer and then soft clip at +/- 32768.

Reply:
Yes I`m aware of this, still in proof of concept stage, will fix this.